packages/context/src/inject.ts

```diff
@@ -60,43 +64,48 @@ export function inject(
   return function markParameterOrPropertyAsInjected(
     // tslint:disable-next-line:no-any
     target: any,
-    propertyKey?: string | symbol,
-    propertyDescriptorOrParameterIndex?:
+    propertyKey: string | symbol,
+    methodDescriptorOrParameterIndex?:
       | TypedPropertyDescriptor<BoundValue>
       | number,
   ) {
-    if (typeof propertyDescriptorOrParameterIndex === 'number') {
+    if (typeof methodDescriptorOrParameterIndex === 'number') {
       // The decorator is applied to a method parameter
       // Please note propertyKey is `undefined` for constructor
-      const injectedArgs: Injection[] =
-        Reflector.getOwnMetadata(PARAMETERS_KEY, target, propertyKey!) || [];
-      injectedArgs[propertyDescriptorOrParameterIndex] = {
-        bindingKey,
-        metadata,
-        resolve,
-      };
-      Reflector.defineMetadata(
+      const paramDecorator: ParameterDecorator = ParameterDecoratorFactory.createDecorator(
         PARAMETERS_KEY,
-        injectedArgs,
-        target,
-        propertyKey!,
+        {
+          bindingKey,
+          metadata,
+          resolve,
+        },
       );
+      paramDecorator(target, propertyKey!, methodDescriptorOrParameterIndex);
     } else if (propertyKey) {
+      // Property or method
       if (typeof Object.getPrototypeOf(target) === 'function') {
         const prop = target.name + '.' + propertyKey.toString();
         throw new Error(
           '@inject is not supported for a static property: ' + prop,
         );
       }
-      // The decorator is applied to a property
-      const injections: {[p: string]: Injection} =
-        Reflector.getOwnMetadata(PROPERTIES_KEY, target) || {};
-      injections[propertyKey] = {bindingKey, metadata, resolve};
-      Reflector.defineMetadata(PROPERTIES_KEY, injections, target);
-    } else {
-      throw new Error(
-        '@inject can only be used on properties or method parameters.',
+      if (methodDescriptorOrParameterIndex) {
+        // Method
+        throw new Error(
+          '@inject cannot be used on a method: ' + propertyKey.toString(),
+        );
+      }
+      const propDecorator: PropertyDecorator = PropertyDecoratorFactory.createDecorator(
+        PROPERTIES_KEY,
+        {
+          bindingKey,
+          metadata,
+          resolve,
+        },
       );
+      propDecorator(target, propertyKey!);
+    } else {
+      // It won't happen here as `@inject` is not compatible with ClassDecorator
```

Comment:
Please throw an error here. While this code may not be called from a TypeScript codebase, I believe it's still possible to call this decorator manually from JavaScript code base. Even if it wasn't, it still a good practice to throw an error if the program ever reaches a code/branch that we did not anticipate.